feat: add MySQL command support

enbiyagoral opened the pull request:
This PR adds MySQL connection support to the Boundary CLI. Users can now securely and easily connect to MySQL targets using the `boundary connect mysql` command.

### What’s Added?
- `boundary connect mysql` command
- Credential and target support for MySQL
- Required tests and documentation updates

## Related Issue
Closes #5740

## Testing
- [x] All tests passed successfully
- [x] MySQL connection tests were performed

## ScreenShot
<img width="754" alt="image" src="https://github.com/user-attachments/assets/1f86587f-f404-4f00-8937-a77efda96e99" />

---
Note: This PR follows the same architecture and UX as the existing helper command for PostgreSQL.

bgajjala8 commented on internal/cmd/commands/connect/mysql.go:
```diff
@@ -0,0 +1,114 @@
+			if err := passfile.Close(); err != nil {
+				return nil, nil, proxy.Credentials{}, fmt.Errorf("Error closing password file after writing to %s: %w", passfile.Name(), err)
+			}
+			args = append(args, "-p"+password)
```
(Blocking) It looks like the password is being set explicitly here, but there's a passfile created earlier. Based on previous patterns, it might be better to use `--defaults-file=# (which reads default options from the given file)` and pass in the passfile that way.  You can read more at `mysql --help`

Quick note: This has to be the first argument, so you might need to adjust how the arguments are being built here.

CC: @RyanDerr @johanbrandhorst

enbiyagoral replied:
✅ Fixed! Now using --defaults-file=%s instead of -p flag, and moved it to be the first argument as recommended

RyanDerr commented on internal/cmd/commands/connect/mysql.go:
```diff
@@ -0,0 +1,114 @@
+			if err := passfile.Close(); err != nil {
+				return nil, nil, proxy.Credentials{}, fmt.Errorf("Error closing password file after writing to %s: %w", passfile.Name(), err)
+			}
+			args = append(args, "-p"+password)
```
Good catch, I think this would arise in the case that credentials from boundary are not injected into the client which would necessitate the `-p` arg, iirc postgres by default prompts for a password if none are provided unlike the `mysql` client.

enbiyagoral replied:
✅ Added else condition to append -p flag when no password is provided, so MySQL will prompt for password input

enbiyagoral commented:
@bgajjala8  @RyanDerr  @johanbrandhorst 

Hi! I've addressed all the review feedback:

1. ✅ Now using `--defaults-file` instead of direct password in args
2. ✅ `--defaults-file` is now the first argument
3. ✅ Added `-p` flag when no password provided for prompt
4. ✅ Followed the same pattern as PostgreSQL implementation

Ready for re-review! Thank you for the guidance.

johanbrandhorst commented on internal/cmd/commands/connect/mysql.go:
```diff
@@ -0,0 +1,119 @@
+				}
+				return nil
+			})
+			_, err = passfile.WriteString(fmt.Sprintf("[client]\npassword=%s", password))
```
I think we could simplify this a little with
```suggestion
			_, err = passfile.Write([]byte("[client]\npassword=" + password))
```

johanbrandhorst commented on internal/cmd/commands/connect/mysql.go:
```diff
@@ -0,0 +1,119 @@
+				return nil, nil, proxy.Credentials{}, fmt.Errorf("Error closing password file after writing to %s: %w", passfile.Name(), err)
+			}
+			// --defaults-file must be the first argument
+			args = append(args, fmt.Sprintf("--defaults-file=%s", passfile.Name()))
```
To make this instruction explicit, consider prepending. This way, even if these instructions are reordered later, this will be passed as the first argument. Note also that `fmt.Sprintf` isn't necessary here either.
```suggestion
			args = append([]string{"--defaults-file=" + passfile.Name()}, args...)
```

johanbrandhorst commented on internal/cmd/commands/connect/mysql.go:
```diff
@@ -0,0 +1,119 @@
+			})
+			_, err = passfile.WriteString(fmt.Sprintf("[client]\npassword=%s", password))
+			if err != nil {
+				return nil, nil, proxy.Credentials{}, fmt.Errorf("Error writing password file to %s: %w", passfile.Name(), err)
```
The passfile is not closed in this case. Consider closing it unconditionally.

```suggestion
				_ = passfile.Close()
				return nil, nil, proxy.Credentials{}, fmt.Errorf("Error writing password file to %s: %w", passfile.Name(), err)
```

johanbrandhorst commented on internal/cmd/commands/connect/mysql.go:
```diff
@@ -0,0 +1,119 @@
+		}
+	}
+	return
+} 
```
Please add a newline to the end of the file.

enbiyagoral replied:
yes, i did.

johanbrandhorst reviewed (changes requested):
This is really cool, thank you! There are a few things we'll need to update before we can merge this:

1. We will need to add some documentation to the website, so users can find this new command. Please add a new file called `mysql.mdx` to https://github.com/hashicorp/boundary/tree/main/website/content/docs/commands/connect where the other connect helpers are documented. You should be able to base it on the postgres one. Also update the `index.mdx` document with references to the new document.
2. We will need to add some tests for this functionality. Fortunately, it's pretty easy to run `mysql` for testing purposes, just as we do with `postgresql`. I think we could add a `StartMysql` helper to https://github.com/hashicorp/boundary/blob/main/testing/internal/e2e/infra/docker.go, then create a test similar to https://github.com/hashicorp/boundary/blob/main/testing/internal/e2e/tests/base_plus/target_tcp_connect_postgres_test.go that will connect to the MySQL instance and run some commands.
3. We'll need to update the CHANGELOG.md with this new feature :). Please follow the existing patterns.

Again, thanks for your contribution to Boundary, this is really exciting!

enbiyagoral commented on testing/internal/e2e/tests/base_plus/target_tcp_connect_mysql_test.go:
```diff
@@ -0,0 +1,89 @@
+	require.Contains(t, buf.String(), "mysql>", "Session did not return expected MySQL prompt")
+	require.Contains(t, buf.String(), c.MysqlDbName, "Session did not return expected output")
+	t.Log("Successfully connected to MySQL target")
+} 
```
Thanks for pointing that out — I missed it. It’s fixed now. 😁

enbiyagoral commented:
Thanks for the feedback! I've updated the test to be completely self-contained - it now starts its own MySQL container using infra.StartMysql(), eliminates any dependency on environment variables, and actually tests the functionality by executing real MySQL commands (SHOW TABLES, SELECT DATABASE()) through the boundary connect mysql command while validating the connection works end-to-end with proper cleanup of all Docker resources.

johanbrandhorst commented on testing/internal/e2e/tests/base_plus/env_test.go:
```diff
@@ -13,6 +13,9 @@ type config struct {
 	PostgresDbName          string `envconfig:"E2E_POSTGRES_DB_NAME" required:"true"`
 	PostgresUser            string `envconfig:"E2E_POSTGRES_USER" required:"true"`
 	PostgresPassword        string `envconfig:"E2E_POSTGRES_PASSWORD" required:"true"`
+	MysqlDbName             string `envconfig:"E2E_MYSQL_DB_NAME" required:"true"`
+	MysqlUser               string `envconfig:"E2E_MYSQL_USER" required:"true"`
+	MysqlPassword           string `envconfig:"E2E_MYSQL_PASSWORD" required:"true"`
```
I don't think this is necessary anymore, please remove.

enbiyagoral replied:
Got it, removed. Resolving

enbiyagoral commented:
> Thanks for your patience. I think the test may fail on some machines because the MySQL output is not accurate. Furthermore, the way our CI is set up at the moment makes it impossible for us to run all the tests against this PR, so I was wondering if I could have your consent for us to recreate this contribution in a PR created by a ~HashiCorp~ IBM employee so the CI will run? You will still have the commit credit of course.

I’m happy to give my consent. Please feel free to proceed, and I appreciate you keeping the commit credit. Let me know if I can assist further.

enbiyagoral commented on testing/internal/e2e/tests/base_plus/target_tcp_connect_mysql_test.go:
Moved it to base as suggested — thanks

johanbrandhorst commented on testing/internal/e2e/tests/base_plus/target_tcp_connect_mysql_test.go:
```diff
@@ -0,0 +1,123 @@
+
+	var buf bytes.Buffer
+	_, _ = io.Copy(&buf, f)
+	require.Contains(t, buf.String(), "mysql>", "Session did not return expected MySQL prompt")
+	require.Contains(t, buf.String(), mysqlDbName, "Session did not return expected output")
```
Did you miss this comment? @enbiyagoral

enbiyagoral replied:
Done, it’s fixed now @johanbrandhorst  😄

johanbrandhorst reviewed (commented):
This looks good to me now! As discussed, lets have @bgajjala8 recreate the PR so we can run CI on it. Thanks for your contribution! We can close this PR out once the other one has been merged.

enbiyagoral commented:
Thanks so much everyone for the help and reviews (@johanbrandhorst, @bgajjala8 )— this was my first contribution to Boundary, and it was a great experience! 🙌 Looking forward to future contributions
